x-pack/plugin/ml-package-loader/src/main/java/org/elasticsearch/xpack/ml/packageloader/action/ModelImporter.java

```diff
@@ -136,17 +146,19 @@ private void uploadVocabulary() throws URISyntaxException {
             true
         );
 
-        executeRequestIfNotCancelled(PutTrainedModelVocabularyAction.INSTANCE, request);
+        executeRequestIfNotCancelled(PutTrainedModelVocabularyAction.INSTANCE, request, listener);
     }
 
     private <Request extends ActionRequest, Response extends ActionResponse> void executeRequestIfNotCancelled(
         ActionType<Response> action,
-        Request request
-    ) {
+        Request request,
+        ActionListener<Response> listener
+    ) throws InterruptedException {
         if (task.isCancelled()) {
             throw new TaskCancelledException(format("task cancelled with reason [%s]", task.getReasonCancelled()));
         }
 
-        client.execute(action, request).actionGet();
+        requestLimiter.acquire();
```

Comment:
We'll still block the thread for the duration of the entire download. Since we know the number of parts, would it be possible to start N downloads concurrently and let the completion listener handles the continuations/failures? That's how we handle throttling in other transport requests. Blocking a thread to wait for the end of a network call is generally not recommended.

Reply:
It is not possible to slice the download into parts for concurrent download as the InputStream does not support that, only 1 thread can read from the InputStream at a time. I've switched to a RefCountingListener to collect the responses which avoids the block waiting for all the writes to finish but I still have a block on acquiring the semaphore when there are too many inflight requests. That is not ideal but it is a vast improvement. I'll get some opinions on how to remove the blocking wait on the semaphore.